cli/parser/options/options.go

```diff
@@ -233,31 +265,209 @@ func DefinitionFrom(info *bfpb.FlagInfo) *Definition {
 	return d
 }
 
-type Option struct {
+// Option represents a single parsed command-line option, including any value
+// that option may have, regardless of if said value was provided with an `=`,
+// as a separate argument, or, in the case of boolean arguments, implicitly via
+// the `--[no]option` syntax. The benefit this interface affords us is largely
+// that we can implement types for specific kinds of options and thus handle
+// the behavior of those options more cleanly and with greater readability by
+// reducing the need for large blocks of branching conditionals.
+type Option interface {
+	arguments.Argument
+	Defined
+	ExpectsValue() bool
+	ClearValue()
+	GetDefinition() *Definition
+	UseShortName(bool)
+	Normalized() Option
+}
+
+// GeneralOption is the concrete implementation of `Option` which supports all
+// types of options and their associated value, if any. It is a stop-gap measure
+// before we move to more specialized `Option` types.
+type GeneralOption struct {
 	*Definition
-	Value string
+	Value         *string
+	UsesShortName bool
+	IsNegative    bool
+	Joined        bool
 }
 
-func (o *Option) AsBool() (bool, error) {
-	switch o.Value {
-	case "yes":
-		return true, nil
-	case "true":
-		return true, nil
-	case "1":
-		return true, nil
-	case "":
+func (o *GeneralOption) GetDefinition() *Definition {
+	return o.Definition
+}
+
+func (o *GeneralOption) ExpectsValue() bool {
+	return o.Definition.RequiresValue() && o.Value == nil
+}
+
+func (o *GeneralOption) GetValue() string {
+	if o.Value != nil {
+		return *o.Value
+	}
+	if o.Definition.HasNegative() {
+		if o.IsNegative {
+			return "0"
+		}
+		return "1"
+	}
+	return ""
+}
+
+func (o *GeneralOption) ClearValue() {
+	o.Value = nil
+	o.IsNegative = false
+}
+
+func (o *GeneralOption) SetValue(value string) {
+	o.Value = &value
+	o.IsNegative = false
+}
+
+func (o *GeneralOption) Format() []string {
+	if o.Value != nil {
+		switch {
+		case o.Definition.PluginID() == StarlarkBuiltinPluginID && o.IsNegative:
+			// Starlark flags can have both a "no" prefix and a value; account for
+			// that here.
+			return []string{"--no" + o.Definition.Name() + "=" + *o.Value}
+		case o.Joined || o.Definition.HasNegative():
+			return []string{"--" + o.Definition.Name() + "=" + *o.Value}
+		case o.UsesShortName && o.Definition.ShortName() != "":
```

Comment:
some more basic questions :sweat_smile: if I'm following this correctly it seems like we're trying to explicitly model the way that the argument was explicitly passed at the command line, whereas before we would canonicalize things early on so we could assume everything would be in canonical form, which marks both parsing and formatting easier. Is the reason we're doing that because of the desire to handle unrecognized plugin args which may/may not be in short-form? (If yes, I wonder if we could simplify by assuming that plugin args will not use short-form, at least until we have a good solution for plugins to define their own option schemas)

Reply:
There are two reasons for this: one is to make sure that when we guess whether or not a plugin arg takes a value, we have as little impact as possible if we get it wrong; the other is that this makes the parsing code much easier to test, since we can test that what we input matches the output of `Format` after parsing exactly, instead of manually partially canonicalizing it. I suppose an alternative to the booleans tracking representation would just be to have every option store its CLI representation, and then `Normalized` could return a copy with that field filled with the canonical representation?

Reply:
`Normalized` does return the canonicalized form; I was suggesting that if we stored the CLI representation, which we currently don't, `Normalized` could replace that representation with the canonicalized representation of the option.